Make fitting more robust and add parallelization

maarzt opened the pull request:
This pull request fixes some issues when running the ellipsoid fitting:
(It replaces the similar PR #3)

* The fitting continues in cases where individual ellipsoid could not be fitted
* The fitting for multiple ellipsoid is run in parallel now
* Logic that may have resulted in endless loops is removed.
* A test for the ellipsoid fitting was added
* Refactoring, very large method have been split into smaller methods that are easier to understand

Commit:
Allow ellipsoid fitting to finish also in cases some ellipsoid could not be fitted at all

Commit:
Add parallelization to ellipsoid fitting and avoid endless loop while computing candidate ellipsoids

Commit:
Avoid endless loop in case less edgels than required points/edgels for ellipsoid fitting

Commit:
Update mastodon version number to be used for ellipsoid fitting plugin

Commit:
Add write lock while updating ellipsoid covariances

Commit:
Add todo regarding num samples parameter

Commit:
Comment out some log statements and now remove unused code to enforce a given number of samplings (was commented out, since i may have lead to endless loops)

maarzt committed:
FitEllipsoidPlugin: Refactor the multi threading code

maarzt committed:
FitEllipsoidPlugin: extract the debugging code into separate methods

maarzt committed:
FitEllipsoidPlugin: Refactoring "processSpot" method into smaller methods

maarzt committed:
FitEllipsoidPlugin: Refactor the counting of fitted ellipsoids

The code for counting found and not fount ellipsoids is now in one method.

maarzt committed:
ArtificialData: make the data for testing reproducible

Using a seeded random number generator with a fixed seed means that we
generate the same test data for each test run. Which allows reproducible
test results.

maarzt committed:
Simplify the SampleEllipsoidEgle.sample(...) method

Subdivide the big method into smaller methods.

Use TIntArrayList as it avoids unnecessary boxing and unboxing of
Integers and thereby puts less pressure on the garbage collector.

maarzt committed:
Improve performance by introducing "numCandidates" parameter

Previously "numSamples" was increased from 100 to 1000 to make sure
that we get at least one ellipsoid candidate if many sampling attempts
failed. But this also slows down the ellipsoid fitting if we already
got enough candidates.

Introduce "numCandidates" restores the original performance, by
stopping the sampling as soon as enough candidates are found.

maarzt committed:
FlipEllipsoidPlugin: don't extend the image unnecessarily

maarzt committed:
FitEllipsoidEdgel: improve error handling

Fitting an ellipsoid might produce unexpected exceptions. Such an exception
is now printed and the fitting continues for the other spots.

Also error handling is simplified, by throwing a NoEllipsoidFoundException
rather than returning null.

maarzt committed:
FitElliposoidPlugin: Refactoring, fix generics used for SourceAndConverter

maarzt committed:
Add test for the FitEllipsoidPlugin

maarzt committed:
POM: don't depend on a snapshot release

maarzt committed:
FitEllipsoidPlugin: refactoring, use the known RealType

maarzt committed:
SampleEllipsoidEdgel: refactor the code to improve readability

maarzt committed:
NoEllipsoidFoundException: add missing license header

maarzt committed:
FitEllipsoidPluginTest: raise the test criteria

maarzt committed:
FitEllipsoidPlugin: remove unnecessary "appModel" parameters

maarzt commented:
@stefanhahmann  Could you please review this PR.
_(Github won't allow me to request you as a reviewer.)_

maarzt commented:
Here is a summary of the changes made in this PR:

**Changes noticeable to our users:**
  - Ellipsoid fitting uses multi threading and is thereby faster.
  - The elliposoid fitting doesn't break down as often.
  - A summary of failed and successfull ellipsoid fittings is shown.
  - More ellipsoid are found (because numSamples was change from 100 to 1000). Fitting quality might even improve.
  
**Changes to the code**
* Method FitEllipoisdPlugin.process(...):
  - The very large method was split into smaller methods.
  - Uses multi threading to speed up the ellipsoid fitting.
  - Continue even if ellipsoid fitting failes for some spot.
  - Count numbers of successfull and failed ellipsoid fitting attempts and print a summary.
* Method SampleEllipsoidEdgel.sample(...):
  - The very large method was split into smaller methods.
  - Throws NoEllipsoidFoundException rather than returning "null".
  - Value for parameter numSamples was changed form 100 to 1000. This increases the chance of finding an ellipsoid. But increases runtime.
  - New parameter numCandidates = 100 introduced. New breaking condition that stop the search for a best fitted ellipsoid candidate after 100 potential candidates were found.
* Add a test FitEllipsoidPluginTest:
  - It runs the FitEllipsoidPlugin on artificial data, and compares the resulting Ellipsoids to the known ideal results.
  - The artifivial data is generated by rendering many multivariant-normal-distribution-functions into on image.

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -162,30 +171,119 @@ private void updateEnabledActions()
-			process( ( SourceAndConverter ) source );
-
-//			System.out.println( "fitSelectedVertices()" );
+			process( Cast.unchecked( source ) );
```
Nice to know that this exists in imglib2

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -162,30 +171,119 @@ private void updateEnabledActions()
 	}
 
 	private static final boolean DEBUG = false;
 
```
```suggestion
private static final boolean DEBUG_UI = false;
```
During testing with a large dataset, I figured out that it is handy to have a separate switch for debugging with/without UI.

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -162,30 +171,119 @@ private void updateEnabledActions()
+			process( Cast.unchecked( source ) );
 		}
 	}
 
```
```suggestion
private static final boolean TRACE = false;
```
While debugging with a large dataset, I found it handy to have different granularities of logging. Of course, it would be better, to introduce a logging framework here, but I would see this not as the purpose of this pull request.

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -162,30 +171,119 @@ private void updateEnabledActions()
+					writeLock.unlock();
+				}
+				found.getAndIncrement();
+				if ( DEBUG )
```
```suggestion
				if ( TRACE )
```

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -195,131 +293,141 @@ private < T extends RealType< T > > void process( final SourceAndConverter< T >
+				angleCutoffDistance,
+				maxCenterDistance );
+
+		if ( DEBUG )
```
```suggestion
		if ( DEBUG_UI )
```

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java:
```diff
@@ -60,91 +64,119 @@ public static Map< Edgel, Double > getCosts(
+		if ( ellipsoid == null )
+			return false;
+
+		// skip degenerate samples
```
```suggestion
		// degenerate ellipsoid
```
This comment could be improved, since it is not inside a loop anymore.

maarzt replied:
The comment was kind of wrong in the first place. Degenerate means that there is a raduis radii[i] == 0. But this was not tested. I added a method to the ellipsoid class: ellipsoid.isLegitimate() that tests is all radii are positive. Using this method makes the comment obsolete.

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java:
```diff
@@ -60,91 +64,119 @@ public static Map< Edgel, Double > getCosts(
+		final double[] radii = ellipsoid.getRadii();
+		if ( Double.isNaN( radii[ 0 ] ) || Double.isNaN( radii[ 1 ] ) || Double.isNaN( radii[ 2 ] ) )
+			return false;
+
```
```suggestion

		// ellipsoid with negative radii
		if ( radii[ 0 ] <= 0 || radii[ 1 ] <= 0 || radii[ 2 ] <= 0 )
			return false;
```
One could add this validation check so that only ellipsoid with positive-only radii are considered valid.

maarzt replied:
I went for a slightly different solution. I added a method `ellipsoid.isLegitimate()` that tests if all the radii of  the ellipsoid are strictly positive. I used this method in the if condition.

stefanhahmann replied:
Okay. Please resolve.

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java:
```diff
@@ -60,91 +64,119 @@ public static Map< Edgel, Double > getCosts(
+		{
+			// refined ellipsoid
+			return fitToInliers( bestEllipsoid, edgels, outsideCutoffDistance, insideCutoffDistance,
+					angleCutoffDistance );
```
```suggestion
			// refined ellipsoid
			Ellipsoid refinedEllipsoid = fitToInliers( edgels, bestEllipsoid, costFunction );

			// test, if refined ellipsoid is valid
			if ( isEllipsoidValid( refinedEllipsoid, expectedCenter, maxCenterDistance, center ) )
				return refinedEllipsoid;

			// return best ellipsoid without refinement, if refined ellipsoid is not valid
			return bestEllipsoid;
```
While testing with a large dataset, I figure out that it makes sense to test, if the refined ellipsoid is valid, since it may cause problems in subsequent steps (e.g. viewing the ellipsoid in BDV), if it is not valid.

In case, it is not valid, there are two options: not returning an ellipsoid at all, or returning the best ellipsoid without refinement. It is hard to decide, which option is better without testing. I would recommend to return the best ellipsoid without refinement in this case. The testing which option is better could be a new issue in this repo.

maarzt replied:
Good change. But why do you suggests to add all the comments, that say exactly whats written below?

stefanhahmann replied:
You are right. Some of them were really too obvious to be written. It is okay for me as it is now. Please resolve.

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java:
```diff
@@ -60,91 +64,119 @@ public static Map< Edgel, Double > getCosts(
+	private static void sampleCoordinatesFromEdgels( List< Edgel > edgels, double[][] coordinates, Random rand, TIntArrayList indices )
+	{
+		indices.clear();
+		for ( int s = 0; s < coordinates.length; ++s )
```
```suggestion
		for ( int s = 0; s < coordinates.length; s++ )
```
Pre- and Post-increment in the loop definition does not make a difference here - as far as I understand. I would prefer post-increment, since this is the more usual way in Java. Otherwise, one has to think, why it is different here.

maarzt replied:
Let's fix these pre increments in one commit for the whole repo in a separate PR.

stefanhahmann replied:
I create a new issue #6. Please resolve.

stefanhahmann commented on src/test/java/org/mastodon/mamut/fitting/FitEllipsoidPluginDemo.java:
```diff
@@ -0,0 +1,62 @@
+/*-
```
Is there some rule in the Mastodon repos that I am not aware of that licensing info is added to every single class?

maarzt replied:
Yes, there is. The release-version script that we use for cutting releases even enforces this rule. You can also use maven to add the version headers `mvn license:update-file-headers`. I don't know if the rule is explicitly written down some where.

stefanhahmann replied:
Okay please resolve.

stefanhahmann commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -195,131 +293,141 @@ private < T extends RealType< T > > void process( final SourceAndConverter< T >
+		final long[] lMin = new long[ 3 ];
+		final long[] lMax = new long[ 3 ];
+
+		for ( int d = 0; d < 3; ++d )
```
```suggestion
		for ( int d = 0; d < 3; d++ )
```
Post increment is more usual in for loops and thus easier to read?

maarzt replied:
I agree. But lets do this in a different PR, that fixes all the pre increments in the for loops.

stefanhahmann reviewed (changes requested):
I tested the code of this Pull Request on large dataset (~370.000 spots in ~700 timeframes).

The suggested changes improve the ellipsoid fitting:
- The fitting is now faster, since it uses parallel processing
- The fitting does not stop, if some individual spots cannot be fitted
- The unit tests added by this PR increase the trust in the results gained by the fitting.

Just a screenshot of the result gained while testing with a dataset: 
![grafik](https://user-images.githubusercontent.com/10515534/232735966-c1916097-63f6-4b7d-a92e-218efc3474ba.png)

One can see that results are okay in many cases, but one can see some errors as well. However, these errors do not seem to be related to the changes of this PR to me.

maarzt committed:
FitEllipsoidPlugin: add more debugging options

Co-authored-by: Stefan Hahmann <stefanhahmann@users.noreply.github.com>

maarzt committed:
SampleEllipsoidEdgel: make sure the refined ellipsoid is valid too

maarzt committed:
SampleEllipsoidEdgel: only use legitimate ellipsoids with all radii > 0

maarzt committed:
SampleEllipsoidEdgel: refactor the check for number of candidates

Using the pre-increment "++candidates >= numCandidate" is harder to read
and understand than splitting it into two lines of code.

stefanhahmann reviewed (approved):
Okay. No further changes needed.

maarzt commented:
Hello @tinevez,

@stefanhahmann and I agree that this PR is ready to be merged. What do you think?
Would you like to review it yourself, or should I merge it?

tinevez commented:
Great!
Tobias was curious about it. Pinging @tpietzsch

maarzt commented:
There's nothing too exciting about this PR.

Here's a **very short summary**: Fix exception handling. Introduce multithreading. Increase paramter numSamples from 100 to 1000 to find more ellipsoids at the same time introduce new breaking condition numCanditates = 100 to keep similar performance. Some refactoring of large methods. And a unit test.

We also noticed that the ellipsoid fitting doesn't work all to well.
I think we need to try a different approach to get better results.
